### DISCUSSION TITLE
What is the purpose of Add Control Binding in MACROS? Or how do you use it?

### DISCUSSION THREAD

**Discussion by Deer-Xmj** [nickcoutsos/keymap-editor#252] [Q&A] [answered]
![image](https://github.com/user-attachments/assets/aa537e92-0675-43bb-8d9f-a4ca8a34fc29)
![image](https://github.com/user-attachments/assets/46978d37-8e7a-4956-8475-9e4c55be7eec)
![image](https://github.com/user-attachments/assets/5436bdba-ccc4-4d16-beb9-77731893392f)
Even though I've clicked Create, I can't figure out exactly how this feature works, if so, can you show me what it does? Appreciate;

**Comment by nickcoutsos** [comment, marked as answer]
Control bindings give you more control in how the macro executes. In a simple case you can use `&macro_press` and `&macro_release` to explicitly say when to hold or release keys, or you can use parameters to turn the macro into a repeatable but customizable behavior.

If you want to do some reading I have an article about [using parameterized macros to re-create an "autoshift"](https://github.com/nickcoutsos/keymap-editor/wiki/Autoshift-using-ZMK-behaviors).

And of course ZMK's documentation goes into detail about [macro controls](https://zmk.dev/docs/keymaps/behaviors/macros#macro-controls).

**Reply by Deer-Xmj** [reply]
Thank you for your reply, I have benefited a lot
